File: c/linked-list/queue.c
#include <stdio.h>
#include <stdlib.h>
#include "queue.h"

Queue* InitQueue() {
	Queue *q = malloc(sizeof(Queue));
	q->head = q->tail = NULL;
	return q;
}

int isEmpty(Queue *q) {
	return q->head == NULL && q->tail == NULL;
}

void enqueue(Queue *q, int n) {
	Node *new_node;
	new_node = malloc(sizeof(Node));	
	new_node->value = n;
	new_node->next = NULL;

	if (q->head == NULL) {
		q->head = q->tail = new_node;
		return;
	} 

	// Queue: head -> [value: 1, next: ->] -> [value: 2, next: NULL]
	// 	tail -> [value: 2, next: NULL]
	q->tail->next = new_node;
	q->tail = new_node;
	// Queue: head -> [value: 1, next: ->] -> [value: 2, next: ->] -> [value: 3, next: NULL]
	//        tail -> [value: 3, next: NULL]

	return;	
}

int dequeue(Queue *q) {
	if (q->head == NULL) return -1;
	Node *tmp = q->head;
	int v = tmp->value;

	q->head = q->head->next;

	if (q->head == NULL) {
		q->tail = NULL;
	}

	free(tmp);
	return v;
}

// int main() {
// 	Queue *queue = InitQueue();
// 	enqueue(queue, 7); // head -> [7 | ] / tail -> [7| ]
// 	enqueue(queue, 3); // head -> [7| ] -> [8| ] / tail -> [8| ]
// 	enqueue(queue, 8); // head -> [7| ] -> [8| ] -> [9| ] / tail -> [9| ]
// 	enqueue(queue, 19); 
// 	enqueue(queue, 1);

// 	printf("====\n");
// 	printf("f %d\n", dequeue(queue));
// 	printf("f %d\n", dequeue(queue));
// 	printf("f %d\n", dequeue(queue));
// 	printf("f %d\n", dequeue(queue));
// 	printf("f %d\n", dequeue(queue));
// 	return 0;
// }

File: kata-machine/src/day1/Queue.ts
type Node<T> = {
    value: T,
    next?: Node<T>,
}

export default class Queue<T> {
    public length: number;
    private head: Node<T> | undefined;
    private tail: Node<T> | undefined

    constructor() {
        this.tail = this.head = undefined;
        this.length = 0;
    }

    enqueue(item: T): void {
        this.length++;
        const node: Node<T> = { value: item };
        if (!this.tail) {
            this.head = this.tail = node;
            return;
        }

        this.tail.next = node;
        this.tail = node;
        return;
    }

    dequeue(): T | undefined {
        if (!this.head) return undefined;

        this.length--;
        
        const head = this.head;
        this.head = this.head?.next

        // free
        head.next = undefined;

        if (this.length === 0) this.head = this.tail = undefined;

        return head.value;
    }

    peek(): T | undefined {
        return this.head?.value;
    }
}